IdentityCore/src/cache/mac/MSIDMacCredentialCacheItem.m

```diff
@@ -0,0 +1,211 @@
+// Copyright (c) Microsoft Corporation.
+// All rights reserved.
+//
+// This code is licensed under the MIT License.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files(the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and / or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions :
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+
+#import "MSIDMacCredentialCacheItem.h"
+
+static NSString *keyDelimiter = @"-";
+
+@interface MSIDMacCredentialCacheItem ()
+
+@property (nonatomic) NSMutableDictionary *cacheObjects;
+@property (nonatomic) dispatch_queue_t queue;
+
+@end
+
+@implementation MSIDMacCredentialCacheItem
+
+- (instancetype)init
+{
+    if(self = [super init])
+    {
+        self.cacheObjects = [NSMutableDictionary dictionary];
+        self.queue = dispatch_queue_create("com.microsoft.universalStorage", DISPATCH_QUEUE_CONCURRENT);
+    }
+    
+    return self;
+}
+
+- (void)setCredential:(MSIDCredentialCacheItem *)token forKey:(NSString *)key
+{
+    dispatch_barrier_async(self.queue, ^{
+        [self.cacheObjects setObject:token forKey:key];
+    });
+}
+
+-(MSIDCredentialCacheItem *)credentialForKey:(NSString *)key
+{
+    __block MSIDCredentialCacheItem *appCredential = nil;
+    
+    dispatch_sync(self.queue, ^{
+        appCredential = [self.cacheObjects objectForKey:key];
+    });
+    
+    return appCredential;
+}
+
+- (void)removeCredentialForKey:(NSString *)key
+{
+    dispatch_barrier_async(self.queue, ^{
+        [self.cacheObjects removeObjectForKey:key];
+    });
+}
+
+- (void)mergeCredential:(MSIDMacCredentialCacheItem *)credential
+{
+    NSArray *keys = [credential allKeys];
+    for (NSString *key in keys)
+    {
+        if (![self credentialForKey:key])
+        {
+            MSIDCredentialCacheItem *item = [credential.cacheObjects objectForKey:key];
```

Comment:
The credentialForKey works on the shared credential . The credential referred to here is the persistent blob retrieved from memory and they are both different. This functions works on merging the in memory blob with the persistent blob. It will be wrong to use credentialForKey method here.